db/redis.go

```diff
@@ -1155,3 +1167,170 @@ func (r *RedisDriver) InsertPaloalto(advs iter.Seq2[models.Paloalto, error]) err
+			if _, ok := newDeps[adv.AdvisoryID]; !ok {
+				newDeps[adv.AdvisoryID] = map[string]map[string]struct{}{}
+			}
+			if _, ok := newDeps[adv.AdvisoryID][adv.CveID]; !ok {
+				newDeps[adv.AdvisoryID][adv.CveID] = map[string]struct{}{}
+			}
+
+			advtoCVEs[adv.AdvisoryID] = append(advtoCVEs[adv.AdvisoryID], adv.CveID)
+
+			for _, product := range adv.Affected {
+				cpePartVendorProductStr := fmt.Sprintf("%s#%s#%s", product.Part, product.Vendor, product.Product)
+				_ = pipe.SAdd(ctx, fmt.Sprintf(cpeKeyFormat, cpePartVendorProductStr), adv.CveID)
+				newDeps[adv.AdvisoryID][adv.CveID][cpePartVendorProductStr] = struct{}{}
+				if _, ok := oldDeps[adv.AdvisoryID]; ok {
+					if _, ok := oldDeps[adv.AdvisoryID][adv.CveID]; ok {
+						delete(oldDeps[adv.AdvisoryID][adv.CveID], cpePartVendorProductStr)
+						if len(oldDeps[adv.AdvisoryID][adv.CveID]) == 0 {
+							delete(oldDeps[adv.AdvisoryID], adv.CveID)
+						}
+					}
+				}
+			}
+			if _, ok := oldDeps[adv.AdvisoryID]; ok {
+				if len(oldDeps[adv.AdvisoryID]) == 0 {
+					delete(oldDeps, adv.AdvisoryID)
+				}
+			}
+		}
+		if _, err = pipe.Exec(ctx); err != nil {
+			return xerrors.Errorf("Failed to exec pipeline. err: %w", err)
+		}
+		bar.Add(len(as))
+	}
+	bar.Finish()
+
+	log.Infof("Refreshed Vulnerabilities.")
+
+	log.Infof("Insert Advisories...")
+	bar = pb.ProgressBarTemplate(`{{cycle . "[                    ]" "[=>                  ]" "[===>                ]" "[=====>              ]" "[======>             ]" "[========>           ]" "[==========>         ]" "[============>       ]" "[==============>     ]" "[================>   ]" "[==================> ]" "[===================>]"}} {{counters .}} files processed. ({{speed .}})`).New(0).Start().SetWriter(func() io.Writer {
+		if viper.GetBool("log-json") {
+			return io.Discard
+		}
+		return os.Stderr
+	}())
+	keys := slices.Collect(maps.Keys(advtoCVEs))
+	for chunk := range slices.Chunk(keys, batchSize) {
+		pipe := r.conn.Pipeline()
+		for _, adv := range chunk {
+			var aj []byte
+			if aj, err = json.Marshal(advtoCVEs[adv]); err != nil {
+				return xerrors.Errorf("Failed to marshal json. err: %w", err)
+			}
+
+			_ = pipe.HSet(ctx, fmt.Sprintf(advKeyFormat, models.CiscoType), adv, string(aj))
+		}
+		if _, err = pipe.Exec(ctx); err != nil {
+			return xerrors.Errorf("Failed to exec pipeline. err: %w", err)
+		}
+		bar.Add(len(chunk))
+	}
+	bar.Finish()
+	log.Infof("Refreshed Advisories.")
+
+	pipe := r.conn.Pipeline()
+	for advID, cves := range oldDeps {
+		if _, ok := newDeps[advID]; !ok {
+			_ = pipe.HDel(ctx, fmt.Sprintf(advKeyFormat, models.CiscoType), advID)
+		}
+		for cveID, cpes := range cves {
+			for cpePartVendorProductStr := range cpes {
+				_ = pipe.SRem(ctx, fmt.Sprintf(cpeKeyFormat, cpePartVendorProductStr), cveID)
+			}
+			if _, ok := newDeps[advID]; !ok {
+				if _, ok := newDeps[advID][cveID]; !ok {
+					_ = pipe.HDel(ctx, fmt.Sprintf(cveKeyFormat, cveID), advID)
+				}
+			}
```

Comment:
```suggestion
			if _, ok := newDeps[advID]; ok {
				if _, ok := newDeps[advID][cveID]; !ok {
					_ = pipe.HDel(ctx, fmt.Sprintf(cveKeyFormat, cveID), advID)
				}
			} else {
				_ = pipe.HDel(ctx, fmt.Sprintf(cveKeyFormat, cveID), advID)
			}
```

I'm not confident of this change... 
The way of my thinking is:
- if advID is in newDeps, we should check the existance of cveID under it.
- if advID is not in newDeps, the advID does not exist any more, better to remove the information about it.
Rubber-ducking may be necessary?

Reply:
fixed: https://github.com/vulsio/go-cve-dictionary/compare/828ecb79b3241b9639f321dbf8d1fce982161cdd..8d39314f9c9f044fc5893e67d32f2a4f4cd8e5b4